python/taichi/lang/matrix.py

```diff
@@ -144,16 +147,49 @@ def assign(self, other):
         for i in range(self.n * self.m):
             self.entries[i].assign(other.entries[i])
 
+    def __add__(self, other):
+        assert self.m == other.m and self.n == other.n, f"Dimension mismatch between shapes ({self.n}, {self.m}), ({other.n}, {other.m})"
+
+        import taichi as ti
+        if impl.inside_kernel():
+            return ti.add(self, other)
+        else:
+            # implement Matrix add in Python scope
+            ret = Matrix(self.n, self.m, empty=True)
+            ret.entries = (np.array(self.entries).reshape(
+                (self.n, self.m)) + np.array(other.entries).reshape(
+                    (other.n, other.m))).flatten().tolist()
+            return ret
+
+    def __sub__(self, other):
+        assert self.m == other.m and self.n == other.n, f"Dimension mismatch between shapes ({self.n}, {self.m}), ({other.n}, {other.m})"
+
+        import taichi as ti
+        if impl.inside_kernel():
+            return ti.sub(self, other)
+        else:
+            # implement Matrix subtract in Python scope
+            ret = Matrix(self.n, self.m, empty=True)
+            ret.entries = (np.array(self.entries).reshape(
+                (self.n, self.m)) - np.array(other.entries).reshape(
+                    (other.n, other.m))).flatten().tolist()
+            return ret
+
```

Comment:
The above 2 are essentially doing very similar things, separating them here to override the operator overloading is just for clarity. Ideally, since now we could get the real `__name__` of the `foo`  from `ops.py/binary` instead of `lambda`, I'm thinking maybe we could move these to `def element_wise_binary` and cosolidate to something like:
```python
if foo.__name__ in {"add", "sub"}: 
    if "add": do_add()
    if "sub": do_sub()
else:
    raise NotImplementedError
```
not sure if that's too far.

(It seems foo cannot be applied on self and other directly unless we intercept the `binary` decorator in `ops.py` based on scope, which looks anti-patten to me...)

Reply:
A quick respond: please try to reuse `python/taichi/lang/common_ops.py`.

Reply:
Sorry, could you help clarify what does it mean to reuse the `common_ops.py`? I thought the `TaichiOperations` base class in that module is shared by `Matrix` and `Expr` so changing its behavior would affect Expr. Also `ti.add` seems to be using the `m.def("expr_add", expr_add)` cbinding  already, so might be too late to intercept there?

Reply:
Sorry for the confusion, it also took me some time to dig out what was the **The above 2** in the original context as Github does not show the right lines after force-pushing ...

As it's shown in this [commit](https://github.com/taichi-dev/taichi/pull/1051/commits/fa8bd56295c11ad8224457d6352753ac45e831e5) from line 151 to 177, the **above 2** originally meant: in the first iteration, I decided to override `__add__` and `__sub__` in `Matrix` class even though these operator overloadings were already defined in the super class `TaichiOperations`, in order to intercept them in Python scope early.

Now since we went down the approach to use `is_pyconstant` directly in the super class in `common_ops.py`, maybe there's no reason to explicitly override them in `Matrix`.